File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.swift
class Solution {
    func maxSum(_ nums: [Int], _ k: Int) -> Int {
        
        let n = nums.count
        
        var ans = Int.min
        
        if k == 0 || n == 1 {
            var sm = 0
            
            for num in nums {
                sm += num
                ans = max(ans, sm)
                
                if sm < 0 {
                    sm = 0
                }
            }
            
            return ans
        }
        
        var cnt = 0
        var cur = 0
        var pref = Array(repeating: 0, count: n + 1)
        
        for i in 0..<n {
            if nums[i] >= 0 {
                cur += nums[i]
            } else {
                cnt += 1
            }
            
            pref[i + 1] = pref[i] + nums[i]
            ans = max(ans, nums[i])
        }
        
        if cnt <= k {
            return cur
        }
        
        var dp = Array(repeating: Array(repeating: 0, count: n), count: n)
        
        for i in 0..<n {
            var pq = [Int]()
            var sm = 0
            
            for j in i..<n {
                if nums[j] >= 0 {
                    dp[i][j] = sm
                    continue
                }
                
                if pq.count < k {
                    pq.append(nums[j])
                    pq.sort(by: >)
                    sm += nums[j]
                } else if pq[0] > nums[j] {
                    sm -= pq[0]
                    pq[0] = nums[j]
                    pq.sort(by: >)
                    sm += nums[j]
                }
            }
        }
        
        for i in 0..<n {
            var pqMax = [Int]()
            var sm = 0
            
            for j in 0..<i {
                if nums[j] < 0 {
                    continue
                }
                
                if pqMax.count < k {
                    pqMax.append(nums[j])
                    pqMax.sort()
                    sm += nums[j]
                } else if pqMax[0] < nums[j] {
                    sm -= pqMax[0]
                    pqMax[0] = nums[j]
                    pqMax.sort()
                    sm += nums[j]
                }
            }
            
            if i + 1 < n {
                for j in stride(from: n - 1, through: i + 1, by: -1) {
                    var curSum = pref[j + 1] - pref[i]
                    curSum -= dp[i][j]
                    curSum += sm
                    
                    ans = max(ans, curSum)
                    
                    if nums[j] < 0 {
                        continue
                    }
                    
                    if pqMax.count < k {
                        pqMax.append(nums[j])
                        pqMax.sort()
                        sm += nums[j]
                    } else if pqMax[0] < nums[j] {
                        sm -= pqMax[0]
                        pqMax[0] = nums[j]
                        pqMax.sort()
                        sm += nums[j]
                    }
                    
                    if !pqMax.isEmpty {
                        ans = max(ans, sm)
                    }
                }
            }
        }
        
        return ans
    }
}

File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn max_sum(nums: Vec<i32>, k: i32) -> i64 {
        let k = k as usize;
        let n = nums.len();

        let mut ans = i32::MIN as i64;

        if k == 0 || n == 1 {
            let mut sm = 0i64;

            for &num in &nums {
                sm += num as i64;
                ans = ans.max(sm);

                if sm < 0 {
                    sm = 0;
                }
            }

            return ans;
        }

        let mut cnt = 0;
        let mut cur = 0i64;
        let mut pref = vec![0i64; n + 1];

        for i in 0..n {
            if nums[i] >= 0 {
                cur += nums[i] as i64;
            } else {
                cnt += 1;
            }

            pref[i + 1] = pref[i] + nums[i] as i64;
            ans = ans.max(nums[i] as i64);
        }

        if cnt <= k {
            return cur;
        }

        let mut dp = vec![vec![0i64; n]; n];

        for i in 0..n {
            let mut pq = BinaryHeap::<i32>::new();
            let mut sm = 0i64;

            for j in i..n {
                if nums[j] >= 0 {
                    dp[i][j] = sm;
                    continue;
                }

                if pq.len() < k {
                    pq.push(nums[j]);
                    sm += nums[j] as i64;
                } else if let Some(&top) = pq.peek() {
                    if top > nums[j] {
                        sm -= top as i64;
                        pq.pop();
                        pq.push(nums[j]);
                        sm += nums[j] as i64;
                    }
                }
            }
        }

        for i in 0..n {
            let mut pqmax = BinaryHeap::<Reverse<i32>>::new();
            let mut sm = 0i64;

            for j in 0..i {
                if nums[j] < 0 {
                    continue;
                }

                if pqmax.len() < k {
                    pqmax.push(Reverse(nums[j]));
                    sm += nums[j] as i64;
                } else if let Some(&Reverse(top)) = pqmax.peek() {
                    if top < nums[j] {
                        sm -= top as i64;
                        pqmax.pop();
                        pqmax.push(Reverse(nums[j]));
                        sm += nums[j] as i64;
                    }
                }
            }

            for j in (i + 1..n).rev() {
                let mut cur_sum = pref[j + 1] - pref[i];
                cur_sum -= dp[i][j];
                cur_sum += sm;

                ans = ans.max(cur_sum);

                if nums[j] < 0 {
                    continue;
                }

                if pqmax.len() < k {
                    pqmax.push(Reverse(nums[j]));
                    sm += nums[j] as i64;
                } else if let Some(&Reverse(top)) = pqmax.peek() {
                    if top < nums[j] {
                        sm -= top as i64;
                        pqmax.pop();
                        pqmax.push(Reverse(nums[j]));
                        sm += nums[j] as i64;
                    }
                }

                if !pqmax.is_empty() {
                    ans = ans.max(sm);
                }
            }
        }

        ans
    }
}